google/cloud/security/scanner/audit/fw_rules_engine.py

```diff
@@ -0,0 +1,199 @@
+
+from collections import namedtuple
+
+from google.cloud.security.common.gcp_type import firewall_rule
+from google.cloud.security.common.util import log_util
+from google.cloud.security.scanner.audit import rules as scanner_rules
+
+
+LOGGER = log_util.get_logger(__name__)
+
+
+# pylint: disable=too-many-instance-attributes
+class Rule(object):
+    """Rule properties from the firewall rules definitions file.
+    Also finds violations.
+    """
+
+    def __init__(self,
+                 rule_name=None,
+                 match_policies=None,
+                 verify_policies=None,
+                 mode=scanner_rules.RuleMode.WHITELIST,
+                 applies_to=scanner_rules.RuleAppliesTo.SELF,
+                 inherit_from_parents=False):
+        """Initialize.
+
+        Args:
+          rule_name (str): The name of the rule.
+          match_policies (list): A list of policy dictionaries.
+          verify_policies (list): A list of policy dictionaries.
+          mode (RuleMode): The RuleMode for this rule.
+          applies_to (RuleAppliesTo): The resources this rule applies to.
+          inherit_from_parents (bool): Whether this rule inherits from parents.
+        """
+        self.name = rule_name
+        self._match_policies = match_policies
+        self._match_rules = None
+        self.mode = mode
+        self._verify_policies = verify_policies
+        self._verify_rules = None
+        self.applies_to = scanner_rules.RuleAppliesTo.verify(applies_to)
+        self.inherit_from_parents = inherit_from_parents
+
+    @staticmethod
+    def create_rules(policies):
+        """Creates FirewallRules from policies.
+
+        Args:
+          policies (list): A list of policy dictionaries.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        match_rules = []
+        for policy in policies:
+            rule = firewall_rule.FirewallRule(**policy)
+            match_rules.append(rule)
+        return match_rules
+
+    @property
+    def match_rules(self):
+        """The FirewallRules used to filter policies.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        if not self._match_rules:
+            self._match_rules = self.create_rules(self._match_policies)
+        return self._match_rules
+
+    @property
+    def verify_rules(self):
+        """The FirewallRules used to check policies.
+
+        Returns:
+          list: A list of FirewallRule.
+        """
+        if not self._verify_rules:
+            self._verify_rules = self.create_rules(self._verify_policies)
+        return self._verify_rules
+
+    def find_policy_violations(self, firewall_policies):
+        """Finds policy violations in a list of firewall policies.
+
+        Args:
+          firewall_policies (list): A list of FirewallRule.
+
+        Yields:
+          iterable: A generator of RuleViolations.
+        """
+        if self.mode == scanner_rules.RuleMode.MATCHES:
```

Comment:
I would move each mode into a separate private method.

Reply:
Something like:
if self.mode == MATCHES:
  violations = self._yield_match_violations(firewall_policies)
elif self.mode == REQUIRED:
  violations = self._yield_required_violations(firewall_policies)
...
\# Return a generator object that yields policy violations
return violations

Reply:
Done.